File: v2/src/main/mongo/monitor.ts
import type { Document, MongoClient } from 'mongodb';
import type { MonitorTick, SlowOp } from '../../shared/monitor.js';

function n(v: unknown): number {
  return typeof v === 'number' ? v : Number(v ?? 0);
}

export async function tick(client: MongoClient): Promise<MonitorTick> {
  const admin = client.db('admin');
  const s = (await admin.command({ serverStatus: 1 })) as Document;

  const ops = (s['opcounters'] as Document) ?? {};
  const network = (s['network'] as Document) ?? {};
  const conns = (s['connections'] as Document) ?? {};
  const mem = (s['mem'] as Document) ?? {};
  const wt = (s['wiredTiger'] as Document | undefined)?.['cache'] as Document | undefined;

  const wtUsed = n(wt?.['bytes currently in the cache']);
  const wtMax = n(wt?.['maximum bytes configured']);

  const opLatencies = ((s['opLatencies'] as Document) ?? {}) as Document;
  const readsLat = ((opLatencies['reads'] as Document) ?? {}) as Document;
  const writesLat = ((opLatencies['writes'] as Document) ?? {}) as Document;
  const cmdsLat = ((opLatencies['commands'] as Document) ?? {}) as Document;

  const globalLock = (s['globalLock'] as Document) ?? {};

  return {
    ts: Date.now(),
    uptimeSeconds: n(s['uptime']),

    opsInsert: n(ops['insert']),
    opsQuery: n(ops['query']),
    opsUpdate: n(ops['update']),
    opsDelete: n(ops['delete']),
    opsGetMore: n(ops['getmore']),
    opsCommand: n(ops['command']),

    networkInBytes: n(network['bytesIn']),
    networkOutBytes: n(network['bytesOut']),

    currentConnections: n(conns['current']),
    availableConnections: n(conns['available']),

    residentMb: n(mem['resident']),
    virtualMb: n(mem['virtual']),
    mappedMb: n(mem['mapped']),

    wtCacheBytesUsed: wtUsed,
    wtCacheBytesMax: wtMax,
    wtCachePercent: wtMax > 0 ? (wtUsed / wtMax) * 100 : 0,

    currentOps: n(((globalLock['currentQueue'] as Document) ?? {})['total']),

    latencyReadsAvgUs:
      n(readsLat['ops']) > 0 ? n(readsLat['latency']) / n(readsLat['ops']) : null,
    latencyWritesAvgUs:
      n(writesLat['ops']) > 0 ? n(writesLat['latency']) / n(writesLat['ops']) : null,
    latencyCommandsAvgUs:
      n(cmdsLat['ops']) > 0 ? n(cmdsLat['latency']) / n(cmdsLat['ops']) : null,
  };
}

export interface ProfilerLevelInput {
  level: 0 | 1 | 2;
  slowMs: number;
}

export async function setProfilerLevel(
  client: MongoClient,
  db: string,
  input: ProfilerLevelInput,
): Promise<void> {
  await client
    .db(db)
    .command({ profile: input.level, slowms: input.slowMs });
}

export async function getProfilerLevel(
  client: MongoClient,
  db: string,
): Promise<ProfilerLevelInput> {
  const r = (await client.db(db).command({ profile: -1 })) as Document;
  return {
    level: ((r['was'] as 0 | 1 | 2) ?? 0) as 0 | 1 | 2,
    slowMs: n(r['slowms']) || 100,
  };
}

export async function listSlowOps(
  client: MongoClient,
  db: string,
  limit = 50,
): Promise<SlowOp[]> {
  try {
    const rows = (await client
      .db(db)
      .collection('system.profile')
      .find({})
      .sort({ ts: -1 })
      .limit(limit)
      .toArray()) as Document[];
    return rows.map((r) => ({
      op: String(r['op'] ?? ''),
      ns: String(r['ns'] ?? ''),
      millis: n(r['millis']),
      ts: r['ts'] ? new Date(String(r['ts'])).getTime() : 0,
      client: r['client'] ? String(r['client']) : undefined,
      user: r['user'] ? String(r['user']) : undefined,
      command: r['command'] ? JSON.stringify(r['command']) : undefined,
    }));
  } catch {
    return [];
  }
}

File: v3/src/main/kotlin/io/mex/mongo/Monitor.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import org.bson.Document

data class MonitorTick(
    val ts: Long,
    val opsInsert: Long, val opsQuery: Long, val opsUpdate: Long,
    val opsDelete: Long, val opsGetMore: Long, val opsCommand: Long,
    val networkInBytes: Long, val networkOutBytes: Long,
    val currentConnections: Long, val availableConnections: Long,
    val residentMb: Long,
    val wtCacheBytesUsed: Long, val wtCacheBytesMax: Long,
    val latencyReadsAvgUs: Double?, val latencyWritesAvgUs: Double?,
) {
    val opsTotal: Long get() = opsInsert + opsQuery + opsUpdate + opsDelete + opsGetMore + opsCommand
    val wtCachePercent: Double get() = if (wtCacheBytesMax == 0L) 0.0 else wtCacheBytesUsed.toDouble() / wtCacheBytesMax * 100.0
}

fun tick(client: MongoClient): MonitorTick {
    val s = client.getDatabase("admin").runCommand(Document("serverStatus", 1))
    val ops = (s["opcounters"] as? Document) ?: Document()
    val network = (s["network"] as? Document) ?: Document()
    val conns = (s["connections"] as? Document) ?: Document()
    val mem = (s["mem"] as? Document) ?: Document()
    val wt = ((s["wiredTiger"] as? Document)?.get("cache") as? Document) ?: Document()
    val ol = (s["opLatencies"] as? Document) ?: Document()

    fun n(doc: Document, key: String) = (doc[key] as? Number)?.toLong() ?: 0L
    fun latencyAvg(name: String): Double? {
        val sub = ol[name] as? Document ?: return null
        val opsCount = (sub["ops"] as? Number)?.toLong() ?: 0L
        val latency = (sub["latency"] as? Number)?.toLong() ?: 0L
        return if (opsCount > 0) latency.toDouble() / opsCount else null
    }

    return MonitorTick(
        ts = System.currentTimeMillis(),
        opsInsert = n(ops, "insert"),
        opsQuery = n(ops, "query"),
        opsUpdate = n(ops, "update"),
        opsDelete = n(ops, "delete"),
        opsGetMore = n(ops, "getmore"),
        opsCommand = n(ops, "command"),
        networkInBytes = n(network, "bytesIn"),
        networkOutBytes = n(network, "bytesOut"),
        currentConnections = n(conns, "current"),
        availableConnections = n(conns, "available"),
        residentMb = n(mem, "resident"),
        wtCacheBytesUsed = (wt["bytes currently in the cache"] as? Number)?.toLong() ?: 0L,
        wtCacheBytesMax = (wt["maximum bytes configured"] as? Number)?.toLong() ?: 0L,
        latencyReadsAvgUs = latencyAvg("reads"),
        latencyWritesAvgUs = latencyAvg("writes"),
    )
}

data class ProfilerLevel(val level: Int, val slowMs: Int)

data class SlowOp(val op: String, val ns: String, val millis: Long, val ts: Long, val command: String?)

fun getProfilerLevel(client: MongoClient, db: String): ProfilerLevel {
    val r = client.getDatabase(db).runCommand(Document("profile", -1))
    return ProfilerLevel(
        level = (r["was"] as? Number)?.toInt() ?: 0,
        slowMs = (r["slowms"] as? Number)?.toInt() ?: 100,
    )
}

fun setProfilerLevel(client: MongoClient, db: String, p: ProfilerLevel) {
    client.getDatabase(db).runCommand(Document("profile", p.level).append("slowms", p.slowMs))
}

fun listSlowOps(client: MongoClient, db: String, limit: Int = 50): List<SlowOp> = runCatching {
    client.getDatabase(db).getCollection("system.profile")
        .find().sort(Document("ts", -1)).limit(limit)
        .into(mutableListOf<Document>())
        .map {
            SlowOp(
                op = it.getString("op") ?: "",
                ns = it.getString("ns") ?: "",
                millis = (it["millis"] as? Number)?.toLong() ?: 0L,
                ts = (it["ts"] as? java.util.Date)?.time ?: 0L,
                command = (it["command"] as? Document)?.toJson(),
            )
        }
}.getOrDefault(emptyList())
